New: padding-line-between-statements (fixes #7356)

mysticatea opened the pull request:
# What is the purpose of this pull request? (put an "X" next to item)

[X] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))

The main issue is; fixes #7356.

Also fixes #7938, fixes #7844, fixes #7568, fixes #6619, and fixes #6070.

# What changes did you make? (Give an overview)

This is working in progress. 

- [x] It requires more test cases.
- [x] Probably I will tweak options.
- [x] I will deprecate some rules.

This PR adds new `newline-between-statements` rule, and deprecates [newline-after-vars], [newline-before-return], and [newline-around-directive] rules.
As mentioned at https://github.com/eslint/eslint/issues/7356#issuecomment-279585090, the point is that this rule focuses "between 2 statements" rather than "before/after a statement". So it should be conflict free.

In addition, this enhances the check for semicolon-less style then resolves some issues.

## EDIT: 2017-02-28

I tweaked this rule from my first proposal. New proposal is here: [gist#8cdde960](https://gist.github.com/mysticatea/8cdde960fd2ca5de850989611989ae3e).

- `newline-between-statements` focuses on only blank lines rather than both line breaks and blank lines. Then now it has simple 3 type: `"any"`, `"always"`, `"never"`.  
   `max-statements-per-line` rule would handle about line breaks (without blank lines).
- Some new statement types: 
    - `cjs-import` and `cjs-export` are handling about #6070.
    - `case` and `default` are handling about #6619.
    - `multiline-block-like` is handling `single` option of JSCS `*PaddingNewLinesAfterBlocks` rule.
- For more flexible configuration, `STATEMENT_TYPE` accepts an array of statement types. Please see the 2nd example of [gist#8cdde960](https://gist.github.com/mysticatea/8cdde960fd2ca5de850989611989ae3e).

## EDIT: 2017-03-30

I'm sorry for late. At last I got time to work on this.

- I changed the option's form as inspired by https://github.com/eslint/eslint/pull/8099#issuecomment-285104406; now it's an array of objects such as `[{blankline: "always", prev: "*", next: "function"}]`.  
  I think my feeling is different to native English, but I felt `before`/`after` is confusable (https://github.com/eslint/eslint/pull/8099#issuecomment-285292323).
- I changed the rule name as inspired by https://github.com/eslint/eslint/pull/8099#issuecomment-285289640; now it's `padding-line-between-statements`.  
  I think good if this rule locates beside [padded-blocks] in our rule list page.
- I soft deprecated 3 rules, [newline-after-vars], [newline-before-return], and [newline-around-directive], as replaced by this rule.

## About autofix

`newline-between-statements` supports autofix.

### never

It will remove blank lines.
However, it's going to remain the blank lines as is if comments exist at the middle of blank lines.
For example:

```js
foo()

foo()

// middle comment

foo()

// normal comment
foo()
```

Gets fixed to:

```js
foo()
foo()

// middle comment

foo()
// normal comment
foo()
```

### always

It will insert a blank line immediately after the previous statement.

```js
foo()
foo() // trailing comment
foo()
// normal comment such as jsdoc.
foo()
```

Gets fixed to:

```js
foo()

foo() // trailing comment

foo()

// normal comment such as jsdoc.
foo()
```

# Is there anything you'd like reviewers to focus on?

- Is this proper to replace 3 rules; [newline-after-vars], [newline-before-return] and [newline-around-directive]? Or should I separate the deprecation from this PR?
- Is this fit to JSCS rules about padding newlines?
- Are there better rule names?

[newline-after-vars]: http://eslint.org/docs/rules/newline-after-vars
[newline-before-return]: http://eslint.org/docs/rules/newline-before-return
[newline-around-directive]: http://eslint.org/docs/rules/newline-around-directive
[padded-blocks]: http://eslint.org/docs/rules/padded-blocks

TheSavior commented on tests/lib/rules/newline-between-statements.js:
```diff
@@ -0,0 +1,167 @@
+
+const ruleTester = new RuleTester();
+
+ruleTester.run("newline-between-statements", rule, {
```
For compatibility, can you include all of the tests from the JSCS rule with the associated configuration to make them pass?

That will help ensure this rule will support the edge cases that JSCS handled: https://github.com/jscs-dev/node-jscs/blob/master/test/specs/rules/require-padding-newlines-after-blocks.js#L50

mysticatea replied:
Thank you for the suggestion.
I added many tests from JSCS rules.
I'm continuing to add more tests.

mysticatea commented:
I updated this proposal. Please see https://github.com/eslint/eslint/pull/8099#issue-208671213 .

I'm still working on this to add more tests, but I think I have done to implement this rule.

ljharb commented:
If I have it set to "never", does that mean:
```js
foo();
foo();
```
is valid? what about:
```js
foo(); foo();
```
?

What would I set it to to ensure that there's 1 newline between statements (the former example) but not 2?

mysticatea commented:
@ljharb Yes, both example is valid. 

> What would I set it to to ensure that there's 1 newline between statements (the former example) but not 2?

Use combination of `newline-between-statements` and `max-statements-per-line`.

not-an-aardvark commented:
I'm wondering if it would be better to use a list of objects instead of a list of arrays:

```js
rules: {
  newline-between-statements: [
    'error',
    { after: '*', before: 'ReturnStatement', linebreak: 'always' }
  ]
}
```

kaicataldo commented:
I like @not-an-aardvark's suggestion a lot - I think it makes it clearer :+1:

mysticatea commented:
Thank you for the suggestion. But I'm confused by it because my first impression is that `ReturnStatement` *is* before linebreaks. :thinking:

alberto commented:
I think we should avoid `newline` in the rule name, to differentiate blank lines from line termination.

We currently mix both cases (e.g. `newline-after-var` and `newline-before-return` vs  `newline-per-chained-call`) but since this rule will deprecate the first two, I think it's a good time to fix this.

We could use `padded-statements` or `blank-lines-between-statements`.

alberto commented:
@mysticatea oh, there are two possible interpretations:
- what is before and after the blankline.
- where should I put the breakline before and after.

mysticatea commented:
Just I refactored with AST selectors. It's the great feature.

mysticatea commented:
I'm sorry for late. At last I got time to work on this.

- I changed the option's form as inspired by https://github.com/eslint/eslint/pull/8099#issuecomment-285104406; now it's an array of objects such as `[{blankline: "always", prev: "*", next: "function"}]`.  
  I think my feeling is different to native English, but I felt `before`/`after` is confusable (https://github.com/eslint/eslint/pull/8099#issuecomment-285292323).
- I changed the rule name as inspired by https://github.com/eslint/eslint/pull/8099#issuecomment-285289640; now it's `padding-line-between-statements`.  
  I think good if this rule locates beside [padded-blocks] in our rule list page.
- I soft deprecated 3 rules, [newline-after-vars], [newline-before-return], and [newline-around-directive], as replaced by this rule.

I think this is ready to review.

Best regards.

not-an-aardvark commented on docs/rules/lines-around-directive.md:
```diff
@@ -1,5 +1,7 @@
 # require or disallow newlines around directives (lines-around-directive)
 
+This rule was **deprecated** in ESLint v3.19.0 and replaced by the [padding-line-between-statements](padding-line-between-statements.md) rule.
```
This will need to be updated from "ESLint v3.19.0" to the version that this PR gets merged in.

not-an-aardvark commented on docs/rules/padding-line-between-statements.md:
````diff
@@ -0,0 +1,229 @@
+}
+```
+
+- `LINEBREAK_TYPE` is one of them.
````
Nitpick: I think it would be better to word this as "LINEBREAK_TYPE is one of the following:"

not-an-aardvark commented on docs/rules/padding-line-between-statements.md:
```diff
@@ -0,0 +1,229 @@
+
+## Examples
+
+It configures as [newline-before-return].
```
Nitpick: I think it would be better to explain this a bit more. For example:

> This configuration would require blank lines before all `return` statements, like the `newline-before-return` rule.

not-an-aardvark commented on docs/rules/padding-line-between-statements.md:
```diff
@@ -0,0 +1,229 @@
+
+
+[newline-after-var]: http://eslint.org/docs/rules/newline-after-var
+[newline-around-directive]: http://eslint.org/docs/rules/newline-around-directive
```
This link is broken. It should be `lines-around-directive`.

not-an-aardvark commented on docs/rules/padding-line-between-statements.md:
```diff
@@ -0,0 +1,229 @@
+
+This rule does nothing if no configuration.
+
+A configuration is an array which has 3 elements; a linebreak type and 2 kind of statements. For example, `{ blankline: "always", prev: "var", next: "return" }` is meaning "it requires one or more blank lines between a variable declaration and a `return` statement."
```
The first sentence is incorrect because the schema has changed -- a configuration is now an object, not an array.

mysticatea replied:
Oops, good catch.

not-an-aardvark commented on docs/rules/padding-line-between-statements.md:
```diff
@@ -0,0 +1,229 @@
+This rule does nothing if no configuration.
+
+A configuration is an array which has 3 elements; a linebreak type and 2 kind of statements. For example, `{ blankline: "always", prev: "var", next: "return" }` is meaning "it requires one or more blank lines between a variable declaration and a `return` statement."
+You can supply any number of the configuration. If an statement pair matches some configurations, it chooses the last matched configuration.
```
Nitpick: I think it would be better to word this as:

> You can supply any number of configurations. If an statement pair matches multiple configurations, the last matched configuration will be used.

not-an-aardvark commented on lib/rules/padding-line-between-statements.js:
```diff
@@ -0,0 +1,585 @@
+                    ]
+                }
+            },
+            type: "array",
```
I think it might be better to have each configuration be a different option:

```json
{
  "padding-line-between-statements": [
    "error",
    { "prev": "*", "next": "*", "blankline": "always" },
    { "prev": "return", "next": "*", "blankline": "never" }
  ]
}
```

Then it would be consistent with other rules, such as `no-restricted-syntax`, `no-restricted-properties`, and `no-restricted-globals`. Also, if we ever create a way to extend a shareable config's configuration for a rule without overwriting it, it will be easier if our rules have consistent APIs.

not-an-aardvark commented on tests/lib/rules/padding-line-between-statements.js:
```diff
@@ -0,0 +1,4420 @@
+        {
+            code: "function a() {\nif (!b) {\nreturn;\n} else {\nreturn b;\n}\n}",
+            options: [[
+                { 
```
I think it would be better to use constants for the error messages to avoid duplication.

```js
const UNEXPECTED_BLANK = "Unexpected blank lines before this statement.";

// ...

errors: [UNEXPECTED_BLANK]
```

not-an-aardvark commented on lib/rules/padding-line-between-statements.js:
```diff
@@ -0,0 +1,585 @@
+        nextToken &&
+        astUtils.isSemicolonToken(semiToken) &&
+        semiToken.loc.start.line !== prevToken.loc.end.line &&
+        semiToken.loc.end.line === nextToken.loc.start.line
```
Will this work correctly for `EmptyStatement` nodes?

```js
if (foo) {
  bar()
}

;[1, 2, 3].map(foo)
```

It seems like the `}` will be considered the last token of the `;` node, which doesn't seem expected.

mysticatea commented on docs/rules/padding-line-between-statements.md:
```diff
@@ -0,0 +1,229 @@
+# Require or disallow padding lines between statements (padding-line-between-statements)
+
+(fixable) The `--fix` option on the [command line](../user-guide/command-line-interface#fix) automatically fixes problems reported by this rule.
```
Thank you!

mysticatea commented on lib/rules/padding-line-between-statements.js:
```diff
@@ -0,0 +1,585 @@
+// Helpers
+//------------------------------------------------------------------------------
+
+const VALID_PARENT = /^(?:Program|(?:Switch|Block)Statement|SwitchCase)$/;
```
This has `SwitchStatement` also, so I need to separate the check of `SwitchStatement`.

mysticatea commented on lib/rules/padding-line-between-statements.js:
```diff
@@ -0,0 +1,585 @@
+ * @param {SourceCode} sourceCode The source code object to get tokens.
+ * @returns {boolean} `true` if the node is a directive.
+ */
+function isDirective(node, sourceCode) {
```
I thought it's not efficient if I make the array of directive prologue then do linear search.

mysticatea commented:
@not-an-aardvark Thank you for the review! I think I fixed those except about `isDirective`.

not-an-aardvark commented on tests/lib/rules/padding-line-between-statements.js:
```diff
@@ -0,0 +1,4445 @@
+// Tests
+//------------------------------------------------------------------------------
+
+const MESSAGE_NEVER = "Unexpected blank lines before this statement.";
```
Nitpick: This message refers to "blank lines", so it might be confusing when there is only one blank line before the statement. I think it would be slightly better to say "Unexpected blank line before this statement."

mysticatea replied:
Thank you, I'll fix it.
The plural is difficult to me.

ilyavolodin commented:
@mysticatea Could you also update replacement.json file with all of the replaced rules please? And since we are making 3 rules deprecated by this, I think this should be marked as `Breaking`

kaicataldo commented:
Technically not breaking, though, right? Since those rules still exist?

mysticatea commented:
Yeah, this is soft-deprecating, not removing.

ilyavolodin commented:
Ok, fair enough.

kaicataldo commented:
Looks like the license check didn't run on this. Maybe try pushing again?

SavePointSam commented:
Can this be merged, @not-an-aardvark / @TheSavior? I'm quite excited about this update. 😄

vitorbal commented on docs/rules/padding-line-between-statements.md:
```diff
@@ -0,0 +1,235 @@
+[lines-around-directive]: http://eslint.org/docs/rules/lines-around-directive
+[newline-after-var]: http://eslint.org/docs/rules/newline-after-var
+[newline-before-return]: http://eslint.org/docs/rules/newline-before-return
+[requirePaddingNewLineAfterVariableDeclaration]: http://jscs.info/rule/requirePaddingNewLineAfterVariableDeclaration
```
I think these links for the JSCS rules could go above, in the "Compatibility" section.

vitorbal commented on docs/rules/padding-line-between-statements.md:
```diff
@@ -0,0 +1,235 @@
+If you don't want to notify warnings about linebreaks, then it's safe to disable this rule.
+
+
+[lines-around-directive]: http://eslint.org/docs/rules/lines-around-directive
```
I think there's a headline missing above this line:

```markdown
## Related Rules
```

mysticatea replied:
Those are just link declarations, don't display.

mysticatea replied:
```md
[foo] ← this is a link.

↓ declare `[foo]`'s destination.
[foo]: http://foo.example.com/

vitorbal replied:
So sorry! I totally misunderstood these. I didn't notice they were just the links 🙏

vitorbal commented:
@SavePointSam looks like this is *almost* ready to go! please bear with us :)
if you're interested, an additional pair of eyes on the code/docs is always helpful.

SavePointSam commented:
@vitorbal No problem! I just have this saved on my bookmark bar and saw there hadn't been any update in about a week and looked like all the conversations were resolved. I'll try and look over the documentation this weekend sometime!

not-an-aardvark commented on docs/rules/padding-line-between-statements.md:
```diff
@@ -0,0 +1,235 @@
+
+This rule does nothing if no configuration.
+
+A configuration is an object which has 3 properties; `blankline`, `prev` and `next`. For example, `{ blankline: "always", prev: "var", next: "return" }` is meaning "it requires one or more blank lines between a variable declaration and a `return` statement."
```
Can we rename the `blankline` property to something like `blankLine` or `blank-line`?

mysticatea commented:
I updated that option name.

There are many cases that property names are camel cases and string constant values are kebab-cases in my impression.

orgkhnargh commented:
Does this rule allow to enforce a specific number of blank lines between statement? Say, if I want two blank lines before class declaration, can I use this rule to enforce that? Docs [here](https://github.com/eslint/eslint/blob/a93a2f951e6a463945c8c1f9673fa4cf31848d7a/docs/rules/padding-line-between-statements.md) don't mention if this is possible. Seems like it's only possible to ensure whether there as *some* blank lines or no blank lines at all. Can someone please clarify this?

Thank you.

mysticatea commented:
> Does this rule allow to enforce a specific number of blank lines between statement?

No. This rule does not have the option to specify the line count of blank lines.

Zephir77167 commented:
Is there any way to enforce newlines between class methods (e.g. in React components)? I tried with: `'padding-line-between-statements': ['error', { 'blankLine': 'always', 'prev': '*', 'next': 'function' }]` but it doesn't seem to work.

Thanks!

mysticatea commented:
> Is there any way to enforce newlines between class methods (e.g. in React components)?

No. This rule is about blank lines between *statements*. Class methods are those are a part of ClassDeclaration or ClassExpression, but not statements.

Probably `padding-line-between-class-members` or something like is needed for the purpose.

helgeFox commented:
Yeah, funny thing is, if you start at #5949 (because you want lines between class methods) and read through that one you get to #7356, which closes 5949. And then you end up here at this issue, that closes 7356. We might actually have an endless loop on our hands guys.

Zephir77167 commented:
My thoughts exactly. Guess we won't have that (and that won't be _so_ terrible)
